staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go

```diff
@@ -0,0 +1,750 @@
+	traceCond sync.Cond
+
+	// flowSchemas holds the flow schema objects, sorted by increasing
+	// numerical (decreasing logical) matching precedence.  Every
+	// FlowSchema in this slice is immutable.
+	flowSchemas apihelpers.FlowSchemaSequence
+
+	// priorityLevelStates maps the PriorityLevelConfiguration object
+	// name to the state for that level.  Every name referenced from a
+	// member of `flowSchemas` has an entry here.
+	priorityLevelStates map[string]*priorityLevelState
+}
+
+// priorityLevelState holds the state specific to a priority level.
+type priorityLevelState struct {
+	// config holds the configuration after defaulting logic has been applied.
+	// Exempt may be true while there are queues, in the case of a priority
+	// level that recently switched from being non-exempt to exempt and whose
+	// queues are still draining.
+	// If there are queues then their parameters are here.
+	config fctypesv1a1.PriorityLevelConfigurationSpec
+
+	// qsCompleter holds the QueueSetCompleter derived from `config`
+	// and `queues` if config is not exempt, nil otherwise.
+	qsCompleter fq.QueueSetCompleter
+
+	// The QueueSet for this priority level.  This is nil if and only
+	// if the priority level is exempt.
+	queues fq.QueueSet
+
+	// quiescing==true indicates that this priority level should be
+	// removed when its queues have all drained.  May be true only if
+	// queues is non-nil.
+	quiescing bool
+
+	// number of goroutines between Controller::Match and calling the
+	// returned StartFunction
+	numPending int
+}
+
+var _ Controller = (*configController)(nil)
+
+// NewController constructs a new Controller
+func NewController(
+	informerFactory kubeinformers.SharedInformerFactory,
+	flowcontrolClient fcclientv1a1.FlowcontrolV1alpha1Interface,
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
+) Controller {
+	grc := counter.NoOp{}
+	return NewTestableController(
+		informerFactory,
+		flowcontrolClient,
+		serverConcurrencyLimit,
+		requestWaitLimit,
+		fqs.NewQueueSetFactory(&clock.RealClock{}, grc),
+	)
+}
+
+// NewTestableController is extra flexible to facilitate testing
+func NewTestableController(
+	informerFactory kubeinformers.SharedInformerFactory,
+	flowcontrolClient fcclientv1a1.FlowcontrolV1alpha1Interface,
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
+	queueSetFactory fq.QueueSetFactory,
+) Controller {
+	cfgCtl := &configController{
+		queueSetFactory:        queueSetFactory,
+		serverConcurrencyLimit: serverConcurrencyLimit,
+		requestWaitLimit:       requestWaitLimit,
+		flowcontrolClient:      flowcontrolClient,
+		priorityLevelStates:    make(map[string]*priorityLevelState),
+	}
+	cfgCtl.traceCond = *sync.NewCond(&cfgCtl.lock)
+	klog.V(2).Infof("NewTestableController with serverConcurrencyLimit=%d, requestWaitLimit=%s", serverConcurrencyLimit, requestWaitLimit)
+	cfgCtl.initializeConfigController(informerFactory)
+	cfgCtl.digestConfigObjects(nil, nil)
```

Comment:
`cfgCtl.digestConfigObjects(..)` _does_ acquire and release the controller's lock.  The acquire will never block, because `cfgCtl` is freshly allocated and there is no concurrent activity at this point.  There are two reasons why using `cfgCtl.digestConfigObjects()` here is a reasonable choice.

(1) There is not currently an existing function that digests config objects without acquiring the lock.

(2) `cfgCtl` is accessed from multiple goroutines and synchronizing its access here follows the simple advice in https://golang.org/ref/mem#tmp_1 .

In fact, in some other languages it is positively necessary to acquire and release the lock before allowing any concurrent access.

In our case, because the access from other goroutines will "happen after" the constructor, we _could_ create and use a function that digests the config objects without acquiring the lock (really this would be a small matter of factoring that functionality out of `lockAndDigestConfigObjects`).  But that seems to me like unnecessary work to make the code more fragile.

Reply:
Sorry, I meant "should we do this in the background, so as not to block the initialization of the apiserver"?

Reply:
No, we need to initialize the data structure to include the effects of the mandatory objects.  And testing shows that there is no deadlock here.  But I agree that we should not even tempt that fate.  That will be an easy fix, given way digestConfigObjects is already factored.

Reply:
revised and commented